### PR TITLE
feat/persist file id in metadata, use it to delete previous content

### PR DESCRIPTION
**Pull request by rbiseck3**
### Description
Currently the id for each vector entry in pinecone is a randomly generated ID. TO prevent duplicates in the destination index, a step was introduces to persist the record id from the `FileData` object in the metadata and delete using that during upload to get rid of duplicates.

### CODE REVIEW

File: unstructured_ingest/v2/processes/connectors/pinecone.py [hunk 1]
```diff
@@ -121,19 +125,23 @@ def conform_dict(self, element_dict: dict) -> dict:
                 }
             )
 
+        metadata = flatten_dict(
+            pinecone_metadata,
+            separator="-",
+            flatten_lists=True,
+            remove_none=True,
+        )
+        metadata["file_id"] = file_data.identifier
```
Comment (vangheem) [inline unstructured_ingest/v2/processes/connectors/pinecone.py:134]: would it make sense to call this `record_id`?

Whatever we choose, we should be consistent across all connectors.

File: unstructured_ingest/v2/processes/connectors/pinecone.py [hunk 2]
```diff
@@ -208,7 +270,16 @@ def run(self, path: Path, file_data: FileData, **kwargs: Any) -> None:
...
+        # Determine if serverless or pod based index
+        pinecone_client = self.connection_config.get_client()
+        index_description = pinecone_client.describe_index(name=self.connection_config.index_name)
+        if "serverless" in index_description.get("spec"):
+            self.serverless_delete_by_record_id(file_data=file_data)
+        elif "pod" in index_description.get("spec"):
+            self.pod_delete_by_record_id(file_data=file_data)
+        else:
+            raise ValueError(f"unexpected spec type in index description: {index_description}")
+        self.serverless_delete_by_record_id(file_data=file_data)
```
Comment (vangheem) [inline unstructured_ingest/v2/processes/connectors/pinecone.py:282]: is this potentially called twice?

File: unstructured_ingest/v2/processes/connectors/pinecone.py (hunk 1 above)
Comment (rbiseck3) [reply unstructured_ingest/v2/processes/connectors/pinecone.py:134]: Ya I wasn't sure what the best label for this was, we don't want to clash with anything that might already exist in the metadata being flattened. I'm cool with `record_id`.

File: unstructured_ingest/v2/processes/connectors/pinecone.py (hunk 2 above)
Comment (rbiseck3) [reply unstructured_ingest/v2/processes/connectors/pinecone.py:282]: Good catch.

File: unstructured_ingest/v2/processes/connectors/pinecone.py (hunk 1 above)
Comment (rbiseck3) [reply unstructured_ingest/v2/processes/connectors/pinecone.py]: Created it as a constant in a new file that will be reused across other uploaders we add this functionality to.

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,9 +1,10 @@
-## 0.2.2-dev1
+## 0.2.2-dev2
 
 ### Enhancements
 
 * **Remove `overwrite` field** from fsspec and databricks connectors
 * **Added migration for GitLab Source V2**
+* **Persist file id in pinecone metadata, use it to delete previous content to prevent duplicates.**
```
Comment (vangheem) [inline CHANGELOG.md:7]:
```suggestion
* **Persist `record_id` in pinecone metadata, use it to delete previous content to prevent duplicates.**
```

**Review by vangheem — APPROVED** [review]

**Pull request merged**